src/channel-qvm.lisp

```diff
@@ -0,0 +1,204 @@
+    ;; not overwrite, because we want the scratch memory to be intact.
+    (rotatef (amplitudes qvm) (%trial-amplitudes qvm))))
+
+
+(defun requires-swapping-p (qvm)
+  "Does the  QVM require swapping of internal pointers?"
+  (and (not (eq (amplitudes qvm) (original-amplitude-pointer qvm)))
+       #+sbcl (eq ':foreign (sb-introspect:allocation-information
+                             (original-amplitude-pointer qvm)))))
+
+
+(defmethod transition ((qvm channel-qvm) (instr quil:gate-application))
+  ;; If any noise rules are matched by theinstruction INSTR, apply the
+  ;; the kraus operators for that noise rule after applying the gate
+  ;; for this instruction.
+  (let* ((gate (pull-teeth-to-get-a-gate instr))
+         (params (mapcar (lambda (p) (force-parameter p qvm))
+                         (quil:application-parameters instr)))
+         (instr-qubits (mapcar #'quil:qubit-index (quil:application-arguments instr)))
+         (noise-rules (noise-rules (noise-model qvm)))
+         (prepend-matched-rule (find-matching-rule noise-rules instr ':before))
+         (append-matched-rule (find-matching-rule noise-rules instr ':after)))
+    (when prepend-matched-rule
+      (apply-all-kraus-maps qvm instr (operation-elements prepend-matched-rule)))
+    (apply #'apply-gate gate (amplitudes qvm) (apply #'nat-tuple instr-qubits) params)
+    (when append-matched-rule
+      (apply-all-kraus-maps qvm instr (operation-elements append-matched-rule)))
+    (incf (pc qvm))
+    qvm))
+
+
+(defmethod transition :around ((qvm channel-qvm) (instr quil:measurement))
+  (let ((ret-qvm (call-next-method)))
+    (apply-classical-readout-noise ret-qvm instr)
+    ret-qvm))
+
+
+(defun rule-matches-instr-p (rule instr position)
+  "Check if rule is matched by instruction data and the position of the match request."
+  (let* ((noise-predicate (predicate-function (noise-predicate rule)))
+         (noise-position (noise-position (noise-predicate rule))))
+    (and (funcall noise-predicate instr) (eq position noise-position))))
+
+
+(defun find-matching-rule (rules instr position)
+  "Return the first rule that matches INSTR/POSITION in the list RULES."
+  (find-if (lambda (rule) (rule-matches-instr-p rule instr position)) rules))
+
+
+(defmethod apply-all-kraus-maps ((qvm channel-qvm) (instr quil:gate-application) kraus-maps)
+  ;; Apply all the kraus operators to the state of the system.
+  (dolist (kraus-map kraus-maps)
+    (apply-kraus-map qvm instr kraus-map)))
+
+
+(defmethod apply-kraus-map ((qvm channel-qvm) (instr quil:gate-application) kraus-ops)
+  ;; Randomly select one of the kraus operators in KRAUS-OPS to apply
+  ;; for the transition.
+  (let* ((amps (%trial-amplitudes qvm))
+         (r (random 1.0d0))
+         (summed 0.0d0)
+         (logical-qubits (quil:application-arguments instr))
+         (qubits (mapcar #'quil:qubit-index logical-qubits)))
+    (check-type amps quantum-state)
```

Comment:
The following change:
```suggestion
    (declare (type quantum-state amps)
```
will be equivalent to `check-type` (i.e., it will trigger an error) when the `safety` value of the current compiling policy is sufficiently high and will turn into a no-op (i.e., it will do nothing) when the `speed` value of the policy is sufficiently high and `safety` is low.  This function is called on each Kraus map twice per call to `transition`, so I think it is preferable to replace the (unconditional) `check-type` by a type annotation that is selectively enforced by the compiler.